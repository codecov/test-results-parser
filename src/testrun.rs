use pyo3::prelude::*;
use pyo3::types::{PyDict, PyFloat, PyInt, PyList, PyString};
use pyo3::{PyAny, PyResult};
use serde::Serialize;
use serde_json::Value;

use crate::validated_string::ValidatedString;

static FRAMEWORKS: [(&str, Framework); 4] = [
    ("pytest", Framework::Pytest),
    ("vitest", Framework::Vitest),
    ("jest", Framework::Jest),
    ("phpunit", Framework::PHPUnit),
];

static EXTENSIONS: [(&str, Framework); 2] =
    [(".py", Framework::Pytest), (".php", Framework::PHPUnit)];

fn check_substring_before_word_boundary(string: &str, substring: &str) -> bool {
    if let Some((_, suffix)) = string.to_lowercase().split_once(substring) {
        return suffix
            .chars()
            .next()
            .is_none_or(|first_char| !first_char.is_alphanumeric());
    }
    false
}

pub fn check_testsuites_name(testsuites_name: &str) -> Option<Framework> {
    FRAMEWORKS
        .iter()
        .filter_map(|(name, framework)| {
            check_substring_before_word_boundary(testsuites_name, name).then_some(*framework)
        })
        .next()
}

#[derive(Clone, Copy, Debug, Serialize, PartialEq)]
pub enum Outcome {
    Pass,
    Failure,
    Skip,
    Error,
}

impl<'py> IntoPyObject<'py> for Outcome {
    type Target = PyString;
    type Output = Bound<'py, Self::Target>;
    type Error = std::convert::Infallible;

    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, std::convert::Infallible> {
        match self {
            Outcome::Pass => Ok("pass".into_pyobject(py)?),
            Outcome::Failure => Ok("failure".into_pyobject(py)?),
            Outcome::Skip => Ok("skip".into_pyobject(py)?),
            Outcome::Error => Ok("error".into_pyobject(py)?),
        }
    }
}

impl<'py> FromPyObject<'py> for Outcome {
    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
        let s = ob.extract::<&str>()?;
        match s {
            "pass" => Ok(Outcome::Pass),
            "failure" => Ok(Outcome::Failure),
            "skip" => Ok(Outcome::Skip),
            "error" => Ok(Outcome::Error),
            _ => Err(PyErr::new::<pyo3::exceptions::PyValueError, _>(format!(
                "Invalid outcome: {}",
                s
            ))),
        }
    }
}

#[derive(Clone, Copy, Debug, Serialize, PartialEq)]
pub enum Framework {
    Pytest,
    Vitest,
    Jest,
    PHPUnit,
}

impl<'py> IntoPyObject<'py> for Framework {
    type Target = PyString;
    type Output = Bound<'py, Self::Target>;
    type Error = std::convert::Infallible;

    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
        match self {
            Framework::Pytest => Ok("Pytest".into_pyobject(py)?),
            Framework::Vitest => Ok("Vitest".into_pyobject(py)?),
            Framework::Jest => Ok("Jest".into_pyobject(py)?),
            Framework::PHPUnit => Ok("PHPUnit".into_pyobject(py)?),
        }
    }
}

impl<'py> FromPyObject<'py> for Framework {
    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
        let s = ob.extract::<&str>()?;
        match s {
            "Pytest" => Ok(Framework::Pytest),
            "Vitest" => Ok(Framework::Vitest),
            "Jest" => Ok(Framework::Jest),
            "PHPUnit" => Ok(Framework::PHPUnit),
            _ => Err(PyErr::new::<pyo3::exceptions::PyValueError, _>(format!(
                "Invalid outcome: {}",
                s
            ))),
        }
    }
}

/// Wrapper for serde_json::Value to enable PyO3 conversion
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct PropertiesValue(pub Option<Value>);

impl<'py> IntoPyObject<'py> for PropertiesValue {
    type Target = PyAny;
    type Output = Bound<'py, Self::Target>;
    type Error = pyo3::PyErr;

    fn into_pyobject(self, py: Python<'py>) -> Result<Self::Output, Self::Error> {
        match self.0 {
            Some(value) => {
                let converted_result = value_to_pyobject(&value, py)?;
                Ok(converted_result.into_bound(py))
            }
            None => Ok(py.None().into_bound(py)),
        }
    }
}

impl<'py> FromPyObject<'py> for PropertiesValue {
    fn extract_bound(ob: &Bound<'py, PyAny>) -> PyResult<Self> {
        if ob.is_none() {
            return Ok(PropertiesValue(None));
        }

        let value = pyobject_to_value(ob)?;
        Ok(PropertiesValue(Some(value)))
    }
}

fn pyobject_to_value(ob: &Bound<'_, PyAny>) -> PyResult<Value> {
    if ob.is_none() {
        return Ok(Value::Null);
    }

    if let Ok(s) = ob.extract::<&str>() {
        return Ok(Value::String(s.to_string()));
    }

    if let Ok(i) = ob.extract::<i64>() {
        return Ok(Value::Number(i.into()));
    }

    if let Ok(f) = ob.extract::<f64>() {
        return Ok(Value::Number(serde_json::Number::from_f64(f).unwrap()));
    }

    if let Ok(b) = ob.extract::<bool>() {
        return Ok(Value::Bool(b));
    }

    if let Ok(list) = ob.downcast::<PyList>() {
        let values = list
            .iter()
            .map(|item| pyobject_to_value(&item))
            .collect::<PyResult<Vec<Value>>>()?;
        return Ok(Value::Array(values));
    }

    if let Ok(dict) = ob.downcast::<PyDict>() {
        let mut map = serde_json::Map::new();
        for (key, value) in dict.iter() {
            let key_str = key.extract::<&str>()?;
            let value_json = pyobject_to_value(&value)?;
            map.insert(key_str.to_string(), value_json);
        }
        return Ok(Value::Object(map));
    }

    // If we can't convert it, try to convert to string as fallback
    let s = ob.str()?.to_string();
    Ok(Value::String(s))
}

fn value_to_pyobject(value: &Value, py: Python) -> PyResult<PyObject> {
    match value {
        Value::String(s) => Ok(PyString::new(py, s).into()),
        Value::Number(n) => Ok(PyFloat::new(py, n.as_f64().unwrap()).into()),
        // Using PyBool returns a Borrowed reference to 'True' or 'False' (singletons)
        // This messed up the `.into()` conversion. So I'm using PyInt instead.
        Value::Bool(b) => Ok(PyInt::new(py, *b as i64).into()),
        Value::Null => Ok(py.None()),
        Value::Array(a) => {
            let values_as_pyobjects = a
                .iter()
                .map(|v| value_to_pyobject(v, py))
                .collect::<Result<Vec<PyObject>, PyErr>>()?;
            let list = PyList::new(py, values_as_pyobjects)?;
            Ok(list.into())
        }
        Value::Object(o) => {
            let dict = PyDict::new(py);
            for (key, value) in o.iter() {
                dict.set_item(key, value_to_pyobject(value, py)?)?;
            }
            Ok(dict.into())
        }
    }
}
// i can't seem to get  pyo3(from_item_all) to work when IntoPyObject is also being derived
#[derive(IntoPyObject, FromPyObject, Clone, Debug, Serialize, PartialEq)]
pub struct Testrun {
    #[pyo3(item)]
    pub name: ValidatedString,
    #[pyo3(item)]
    pub classname: ValidatedString,
    #[pyo3(item)]
    pub duration: Option<f64>,
    #[pyo3(item)]
    pub outcome: Outcome,
    #[pyo3(item)]
    pub testsuite: ValidatedString,
    #[pyo3(item)]
    pub failure_message: Option<String>,
    #[pyo3(item)]
    pub filename: Option<ValidatedString>,
    #[pyo3(item)]
    pub build_url: Option<String>,
    #[pyo3(item)]
    pub computed_name: ValidatedString,
    #[pyo3(item)]
    pub properties: PropertiesValue,
}

impl Testrun {
    pub fn framework(&self) -> Option<Framework> {
        for (name, framework) in FRAMEWORKS {
            if check_substring_before_word_boundary(&self.testsuite, name) {
                return Some(framework);
            }
        }

        for (extension, framework) in EXTENSIONS {
            if check_substring_before_word_boundary(&self.classname, extension)
                || check_substring_before_word_boundary(&self.name, extension)
            {
                return Some(framework);
            }

            if let Some(message) = &self.failure_message {
                if check_substring_before_word_boundary(message, extension) {
                    return Some(framework);
                }
            }

            if let Some(filename) = &self.filename {
                if check_substring_before_word_boundary(filename, extension) {
                    return Some(framework);
                }
            }
        }
        None
    }
}

#[derive(Clone, Debug, Serialize, IntoPyObject)]
pub struct ParsingInfo {
    pub framework: Option<Framework>,
    pub testruns: Vec<Testrun>,
    pub warnings: Vec<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn setup() {
        pyo3::prepare_freethreaded_python();
    }

    #[test]
    fn test_detect_framework_testsuites_name_no_match() {
        let f = check_testsuites_name("whatever");
        assert_eq!(f, None)
    }

    #[test]
    fn test_detect_framework_testsuites_name_match() {
        let f = check_testsuites_name("jest tests");
        assert_eq!(f, Some(Framework::Jest))
    }

    #[test]
    fn test_detect_framework_testsuite_name() {
        let t = Testrun {
            classname: ValidatedString::default(),
            name: ValidatedString::default(),
            duration: None,
            outcome: Outcome::Pass,
            testsuite: "pytest".try_into().unwrap(),
            failure_message: None,
            filename: None,
            build_url: None,
            computed_name: ValidatedString::default(),
            properties: PropertiesValue(None),
        };
        assert_eq!(t.framework(), Some(Framework::Pytest))
    }

    #[test]
    fn test_detect_framework_filenames() {
        let t = Testrun {
            classname: ValidatedString::default(),
            name: ValidatedString::default(),
            duration: None,
            outcome: Outcome::Pass,
            testsuite: ValidatedString::default(),
            failure_message: None,
            filename: Some(".py".try_into().unwrap()),
            build_url: None,
            computed_name: ValidatedString::default(),
            properties: PropertiesValue(None),
        };
        assert_eq!(t.framework(), Some(Framework::Pytest))
    }

    #[test]
    fn test_detect_framework_example_classname() {
        let t = Testrun {
            classname: ".py".try_into().unwrap(),
            name: ValidatedString::default(),
            duration: None,
            outcome: Outcome::Pass,
            testsuite: ValidatedString::default(),
            failure_message: None,
            filename: None,
            build_url: None,
            computed_name: ValidatedString::default(),
            properties: PropertiesValue(None),
        };
        assert_eq!(t.framework(), Some(Framework::Pytest))
    }

    #[test]
    fn test_detect_framework_example_name() {
        let t = Testrun {
            classname: ValidatedString::default(),
            name: ".py".try_into().unwrap(),
            duration: None,
            outcome: Outcome::Pass,
            testsuite: ValidatedString::default(),
            failure_message: None,
            filename: None,
            build_url: None,
            computed_name: ValidatedString::default(),
            properties: PropertiesValue(None),
        };
        assert_eq!(t.framework(), Some(Framework::Pytest))
    }

    #[test]
    fn test_detect_framework_failure_messages() {
        let t = Testrun {
            classname: ValidatedString::default(),
            name: ValidatedString::default(),
            duration: None,
            outcome: Outcome::Pass,
            testsuite: ValidatedString::default(),
            failure_message: Some(".py".to_string()),
            filename: None,
            build_url: None,
            computed_name: ValidatedString::default(),
            properties: PropertiesValue(None),
        };
        assert_eq!(t.framework(), Some(Framework::Pytest))
    }

    #[test]
    fn test_detect_build_url() {
        let t = Testrun {
            classname: ValidatedString::default(),
            name: ValidatedString::default(),
            duration: None,
            outcome: Outcome::Pass,
            testsuite: ValidatedString::default(),
            failure_message: Some(".py".to_string()),
            filename: None,
            build_url: Some("https://example.com/build_url".to_string()),
            computed_name: ValidatedString::default(),
            properties: PropertiesValue(None),
        };
        assert_eq!(t.framework(), Some(Framework::Pytest))
    }

    #[test]
    fn test_properties_into_none_conversion() {
        setup();
        let property = PropertiesValue(None);
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            assert!(property_py.is_none());
        })
    }

    #[test]
    fn test_properties_into_string_conversion() {
        setup();
        let property = PropertiesValue(Some(json!("test_string")));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            assert_eq!(
                round_trip_value,
                PropertiesValue(Some(json!("test_string")))
            );
        })
    }

    #[test]
    fn test_properties_into_integer_conversion() {
        setup();
        let property = PropertiesValue(Some(json!(42)));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            // The conversion of ANY number into PyObj casts that number into float
            assert_eq!(round_trip_value, PropertiesValue(Some(json!(42.0))));
        })
    }

    #[test]
    fn test_properties_into_boolean_conversion() {
        setup();
        let property = PropertiesValue(Some(json!(true)));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            // Conversion from booleans are cast into integers
            assert_eq!(round_trip_value, PropertiesValue(Some(json!(1))));
        })
    }

    #[test]
    fn test_properties_into_empty_list_conversion() {
        setup();
        let property = PropertiesValue(Some(json!([])));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            assert_eq!(round_trip_value, PropertiesValue(Some(json!([]))));
        })
    }

    #[test]
    fn test_properties_into_list_with_values_conversion() {
        setup();
        let property = PropertiesValue(Some(json!(["item1", 123, 4.20, true])));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            // Note: booleans get converted to integers in the round trip
            assert_eq!(
                round_trip_value,
                PropertiesValue(Some(json!(["item1", 123.0, 4.20, 1])))
            );
        })
    }

    #[test]
    fn test_properties_into_empty_dict_conversion() {
        setup();
        let property = PropertiesValue(Some(json!({})));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            assert_eq!(round_trip_value, PropertiesValue(Some(json!({}))));
        })
    }

    #[test]
    fn test_properties_into_dict_with_values_conversion() {
        setup();
        let property = PropertiesValue(Some(json!({
            "string_key": "string_value",
            "int_key": 456,
            "bool_key": false
        })));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            // Note: booleans get converted to integers in the round trip
            assert_eq!(
                round_trip_value,
                PropertiesValue(Some(json!({
                    "string_key": "string_value",
                    "int_key": 456.0,
                    "bool_key": 0
                })))
            );
        })
    }

    #[test]
    fn test_properties_into_nested_dict_conversion() {
        setup();
        let property = PropertiesValue(Some(json!({
            "outer_key": "outer_value",
            "nested": {
                "inner_key": "inner_value"
            }
        })));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            assert_eq!(
                round_trip_value,
                PropertiesValue(Some(json!({
                    "outer_key": "outer_value",
                    "nested": {
                        "inner_key": "inner_value"
                    }
                })))
            );
        })
    }

    #[test]
    fn test_properties_into_list_with_dict_conversion() {
        setup();
        let property = PropertiesValue(Some(json!([
            "list_item",
            {
                "key": "value"
            }
        ])));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            assert_eq!(
                round_trip_value,
                PropertiesValue(Some(json!([
                    "list_item",
                    {
                        "key": "value"
                    }
                ])))
            );
        })
    }

    #[test]
    fn test_properties_into_dict_with_list_conversion() {
        setup();
        let property = PropertiesValue(Some(json!({
            "numbers": [1, 2, 3],
            "name": "test"
        })));
        Python::with_gil(|py| {
            let property_py = property
                .into_pyobject(py)
                .expect("Failed to convert PropertiesValue to Python object");
            let round_trip_value = PropertiesValue::extract_bound(&property_py)
                .expect("Failed to extract PropertiesValue from Python object");
            assert_eq!(
                round_trip_value,
                PropertiesValue(Some(json!({
                    "numbers": [1.0, 2.0, 3.0],
                    "name": "test"
                })))
            );
        })
    }
}
