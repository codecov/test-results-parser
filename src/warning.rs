use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct WarningInfo {
    pub message: String,
    pub location: u64,
}

impl WarningInfo {
    pub fn new(message: String, location: u64) -> Self {
        Self { message, location }
    }
}
