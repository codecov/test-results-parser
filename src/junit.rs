use pyo3::exceptions::PyException;
use pyo3::prelude::*;

use std::collections::HashMap;

use quick_xml::events::attributes::Attributes;
use quick_xml::events::Event;
use quick_xml::reader::Reader;

use crate::testrun::{Outcome, Testrun};

// from https://gist.github.com/scott-codecov/311c174ecc7de87f7d7c50371c6ef927#file-cobertura-rs-L18-L31
fn attributes_map(attributes: Attributes) -> HashMap<String, String> {
    return attributes
        .filter_map(|a| {
            if let Ok(attr) = a {
                let bytes = attr.value.into_owned();
                let value = String::from_utf8(bytes).unwrap();
                let key = String::from_utf8(attr.key.local_name().as_ref().to_vec()).unwrap();
                Some((key, value))
            } else {
                None
            }
        })
        .collect::<HashMap<_, _>>();
}

fn populate(attr_hm: &HashMap<String, String>, testsuite: String) -> Testrun {
    let name = format!(
        "{}::{}",
        attr_hm.get("classname").unwrap(),
        attr_hm.get("name").unwrap()
    );

    let duration = attr_hm.get("time").unwrap().to_string().parse().unwrap();

    Testrun {
        name,
        duration,
        outcome: Outcome::Pass,
        testsuite: testsuite,
    }
}

#[pyfunction]
pub fn parse_junit_xml(file_bytes: Vec<u8>) -> PyResult<Vec<Testrun>> {
    let file_string = String::from_utf8_lossy(&file_bytes).into_owned();
    let mut reader = Reader::from_str(file_string.as_str());
    reader.trim_text(true);

    let mut list_of_test_runs = Vec::new();

    let mut buf = Vec::new();

    let mut saved_testrun: Option<Testrun> = None;

    let mut curr_testsuite = String::new();

    loop {
        match reader.read_event_into(&mut buf) {
            Err(e) => {
                break Err(PyException::new_err(format!(
                    "Error at position: {} {:?}",
                    reader.buffer_position(),
                    e
                )))
            }
            Ok(Event::Eof) => {
                break Ok(list_of_test_runs);
            }
            Ok(Event::Start(e)) => match e.name().as_ref() {
                b"testcase" => {
                    let attr_hm = attributes_map(e.attributes());
                    saved_testrun = Some(populate(&attr_hm, curr_testsuite.clone()));
                }
                b"skipped" => {
                    let mut testrun = saved_testrun.unwrap();
                    testrun.outcome = Outcome::Skip;
                    saved_testrun = Some(testrun);
                }
                b"error" => {
                    let mut testrun = saved_testrun.unwrap();
                    testrun.outcome = Outcome::Error;
                    saved_testrun = Some(testrun);
                }
                b"failure" => {
                    let mut testrun = saved_testrun.unwrap();
                    testrun.outcome = Outcome::Failure;
                    saved_testrun = Some(testrun);
                }
                b"testsuite" => {
                    let attr_hm = attributes_map(e.attributes());

                    curr_testsuite = attr_hm.get("name").unwrap().to_string();
                }
                _ => (),
            },
            Ok(Event::End(e)) => match e.name().as_ref() {
                b"testcase" => {
                    list_of_test_runs.push(saved_testrun.unwrap());
                    saved_testrun = None;
                }
                _ => (),
            },
            Ok(Event::Empty(e)) => match e.name().as_ref() {
                b"testcase" => {
                    let attr_hm = attributes_map(e.attributes());
                    list_of_test_runs.push(populate(&attr_hm, curr_testsuite.clone()));
                }
                _ => (),
            },
            Ok(Event::Text(_)) => (),

            // There are several other `Event`s we do not consider here
            _ => (),
        }
        buf.clear()
    }
}
