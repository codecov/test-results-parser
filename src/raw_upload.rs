use anyhow::Context;

use base64::prelude::*;
use pyo3::prelude::*;
use std::collections::HashSet;
use std::io::prelude::*;

use flate2::bufread::ZlibDecoder;

use quick_xml::reader::Reader;
use serde::Deserialize;

use crate::junit::{get_position_info, use_reader};
use crate::testrun::ParsingInfo;
use crate::warning::WarningInfo;

#[derive(Deserialize, Debug, Clone)]
struct TestResultFile {
    filename: String,
    data: String,
}
#[derive(Deserialize, Debug, Clone)]
struct RawTestResultUpload {
    #[serde(default)]
    network: Option<HashSet<String>>,
    test_results_files: Vec<TestResultFile>,
}

#[derive(Debug, Clone)]
struct ReadableFile {
    filename: String,
    data: Vec<u8>,
}

const LEGACY_FORMAT_PREFIX: &[u8] = b"# path=";
const LEGACY_FORMAT_SUFFIX: &[u8] = b"<<<<<< EOF";

fn serialize_to_legacy_format(readable_files: Vec<ReadableFile>) -> Vec<u8> {
    let mut res = Vec::new();
    for file in readable_files {
        res.extend_from_slice(LEGACY_FORMAT_PREFIX);
        res.extend_from_slice(file.filename.as_bytes());
        res.extend_from_slice(b"\n");
        res.extend_from_slice(&file.data);
        res.extend_from_slice(b"\n");
        res.extend_from_slice(LEGACY_FORMAT_SUFFIX);
        res.extend_from_slice(b"\n");
    }
    res
}

// the warnings should be ordered by location because they're pushed to the vec as we parse
// so we can guarantee that warning[x].location >= warning[x - 1].location
// implicitly tested by warnings-junit.xml
fn format_warnings(input: &[u8], warnings: Vec<WarningInfo>, filename: &str) -> Vec<String> {
    let mut offset = 0;
    let mut result = Vec::new();
    let mut line = 1;
    let mut col = 0;
    let mut input_iter = input.iter();
    for warning in warnings {
        for bytes in input_iter
            .by_ref()
            .take((warning.location - offset) as usize)
        {
            if *bytes == b'\n' {
                line += 1;
                col = 0;
            } else {
                col += 1;
            }
        }
        offset += warning.location;
        result.push(if warning.message.contains('\n') {
            format!("{}\n\nAt {}:{} in {}", warning.message, line, col, filename)
        } else {
            format!("{} at {}:{} in {}", warning.message, line, col, filename)
        });
    }
    result
}

#[pyfunction]
#[pyo3(signature = (raw_upload_bytes))]
pub fn parse_raw_upload(raw_upload_bytes: &[u8]) -> anyhow::Result<(Vec<ParsingInfo>, Vec<u8>)> {
    let upload: RawTestResultUpload =
        serde_json::from_slice(raw_upload_bytes).context("Error deserializing json")?;
    let network: Option<HashSet<String>> = upload.network;

    let mut results: Vec<ParsingInfo> = Vec::with_capacity(upload.test_results_files.len());
    let mut readable_files: Vec<ReadableFile> = Vec::with_capacity(upload.test_results_files.len());

    for file in upload.test_results_files {
        let decoded_file_bytes = BASE64_STANDARD
            .decode(file.data)
            .context("Error decoding base64")?;

        let mut decoder = ZlibDecoder::new(decoded_file_bytes.as_slice());

        let mut decompressed_file_bytes = Vec::new();
        decoder
            .read_to_end(&mut decompressed_file_bytes)
            .context("Error decompressing file")?;

        let mut reader = Reader::from_reader(decompressed_file_bytes.as_slice());
        reader.config_mut().trim_text(true);
        let (framework, testruns, warnings) = use_reader(&mut reader, network.as_ref())
            .with_context(|| {
                let pos_conversion = reader.buffer_position().try_into();
                match pos_conversion {
                    Ok(pos) => {
                        let (line, col) = get_position_info(&decompressed_file_bytes, pos);
                        format!(
                            "Error parsing JUnit XML in {} at {}:{}",
                            file.filename, line, col
                        )
                    }
                    Err(_) => format!("Error parsing JUnit XML in {}", file.filename),
                }
            })?;

        let warning_strings: Vec<String> =
            format_warnings(&decompressed_file_bytes, warnings, &file.filename);

        let parsing_info = ParsingInfo {
            framework,
            testruns,
            warnings: warning_strings,
        };
        results.push(parsing_info);

        let readable_file = ReadableFile {
            data: decompressed_file_bytes,
            filename: file.filename,
        };
        readable_files.push(readable_file);
    }

    let readable_file = serialize_to_legacy_format(readable_files);

    Ok((results, readable_file))
}

#[cfg(test)]
mod tests {
    use base64::engine::general_purpose::STANDARD as BASE64_STANDARD;
    use base64::Engine;
    use flate2::Compression;
    use std::io::Write;

    use super::*;
    use insta::{assert_yaml_snapshot, glob};

    fn file_into_bytes(filename: &str) -> Vec<u8> {
        let upload = std::fs::read(filename).unwrap();
        let mut encoder = flate2::write::ZlibEncoder::new(Vec::new(), Compression::default());
        encoder.write_all(&upload).unwrap();
        let compressed = encoder.finish().unwrap();
        let base64_data = BASE64_STANDARD.encode(compressed);
        let upload_json = format!(
            r#"{{"network": [], "test_results_files": [{{"filename": "{}", "format": "base64+compressed", "data": "{}"}}]}}"#,
            filename.split('/').next_back().unwrap(),
            base64_data,
        );
        upload_json.into()
    }

    #[test]
    fn test_parse_raw_upload_success() {
        glob!("../tests", "*.xml", |path| {
            let upload_json = file_into_bytes(path.to_str().unwrap());
            let result = parse_raw_upload(&upload_json);
            match result {
                Ok((results, _)) => assert_yaml_snapshot!(results),
                Err(e) => {
                    assert_yaml_snapshot!(e.to_string());
                }
            }
        });
    }
}
