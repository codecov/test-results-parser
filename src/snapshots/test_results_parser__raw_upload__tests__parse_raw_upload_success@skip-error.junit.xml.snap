---
source: src/raw_upload.rs
assertion_line: 173
expression: results
input_file: tests/skip-error.junit.xml
---
- framework: Pytest
  testruns:
    - name: test_subtract
      classname: tests.test_math.TestMath
      duration: 0.1
      outcome: Error
      testsuite: pytest
      failure_message: hello world
      filename: ~
      build_url: ~
      computed_name: "tests.test_math.TestMath::test_subtract"
      properties: ~
    - name: test_multiply
      classname: tests.test_math.TestMath
      duration: 0.1
      outcome: Error
      testsuite: pytest
      failure_message: ~
      filename: ~
      build_url: ~
      computed_name: "tests.test_math.TestMath::test_multiply"
      properties: ~
    - name: test_add
      classname: tests.test_math.TestMath
      duration: 0.1
      outcome: Skip
      testsuite: pytest
      failure_message: ~
      filename: ~
      build_url: ~
      computed_name: "tests.test_math.TestMath::test_add"
      properties: ~
  warnings: []
